gymnasium/wrappers/to_array.py

```
@@ -0,0 +1,256 @@
+
+        Args:
+            env: The Array API compatible environment to wrap
+            env_xp: The Array API framework the environment is on
+            target_xp: The Array API framework to convert to
+            env_device: The device the environment is on
+            target_device: The device on which Arrays should be returned
+        """
+        gym.Wrapper.__init__(self, env)
+
+        self._env_xp = module_namespace(env_xp)
+        self._target_xp = module_namespace(target_xp)
+        self._env_device: Device | None = env_device
+        self._target_device: Device | None = target_device
+
+    def step(
+        self, action: WrapperActType
+    ) -> tuple[WrapperObsType, SupportsFloat, bool, bool, dict]:
+        """Performs the given action within the environment.
+
+        Args:
+            action: The action to perform as any Array API compatible array
+
+        Returns:
+            The next observation, reward, termination, truncation, and extra info
+        """
+        action = to_xp(action, xp=self._env_xp, device=self._env_device)
+        obs, reward, terminated, truncated, info = self.env.step(action)
+
+        return (
+            to_xp(obs, xp=self._target_xp, device=self._target_device),
+            float(reward),
+            bool(terminated),
+            bool(truncated),
+            to_xp(info, xp=self._target_xp, device=self._target_device),
+        )
+
+    def reset(
+        self, *, seed: int | None = None, options: dict[str, Any] | None = None
+    ) -> tuple[WrapperObsType, dict[str, Any]]:
+        """Resets the environment returning observation and info as Array from any Array API compatible framework.
+
+        Args:
+            seed: The seed for resetting the environment
+            options: The options for resetting the environment, these are converted to jax arrays.
+
+        Returns:
+            xp-based observations and info
+        """
+        if options:
```

This is where things might get a bit dicey. Options are meant to be relatively free-form, and I don't think we should force any specific structure on this object. In particular, values might be strings, or any other exotic objects. It's not obvious to me whether they should be automatically converted. Same thing applies to info dicts. Does the current `to_xp` function handle arbitrary objects that might not be easily convertible?

I'd consider adding an extra parameter to this (and the vector version) wrapper, defining the behavior of options/infos/anything else that might be questionable. Else this might cause a lot of weird interactions.

As an example: an option is an int which is then converted to a string, or maybe even directly used as a key in a dictionary. Automatically converting options to arrays might would have some hard to predict consequences.

(side note: I realize that this issue was already present, but might as well fix it/think about it if we're revamping the whole thing anyways)

---

It does not. Currently, anything that can't be converted raises an error. There is also no way to specify that e.g. a numpy array in the info dict should not be converted. The list of objects we _can_ convert is limited to:

- Numbers
- Mappings
- Iterables (-> Arrays)
- None

So even a string should raise an error. We could ofc also return objects that are not supported as they are. Alternatively, one could have `convert_info` and `convert_options` arguments, but I am also in favor of keeping all init args limited and simple.